components/kf-monitoring/MetricsExporter.go

```diff
@@ -0,0 +1,161 @@
+	requestFailureGauge = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "requests_failure_gauge",
+		Help: "Number of requests_failure_gauge",
+	}, []string{COMPONENT, KIND, NAMESPACE, ACTION, SEVERITY})
+
+	// Linear latencies
+	requestLinearLatency = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "request_linear_latency",
+		Help:    "A histogram of request_linear_latency",
+		Buckets: prometheus.LinearBuckets(1, 1, 15),
+	}, []string{COMPONENT, KIND, NAMESPACE, ACTION, SEVERITY})
+
+	// Exponential latencies
+	requestExponentialLatency = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "request_exponential_latency",
+		Help:    "A histogram of request_exponential_latency",
+	}, []string{COMPONENT, KIND, NAMESPACE, ACTION, SEVERITY})
+
+	serviceHeartbeat = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "service_heartbeat",
+		Help: "Heartbeat signal every 10 seconds indicating pods are alive.",
+	})
+)
+
+func init() {
+	// Register prometheus counters
+	prometheus.MustRegister(requestCounter)
+	prometheus.MustRegister(requestFailureCounter)
+	prometheus.MustRegister(requestGauge)
+	prometheus.MustRegister(requestFailureGauge)
+	prometheus.MustRegister(requestLinearLatency)
+	prometheus.MustRegister(requestExponentialLatency)
+	prometheus.MustRegister(serviceHeartbeat)
+}
+
+type MetricsExporter struct {
+	component	string
+	// Add error channel so that kf_monitor user can get notified when monitoring thread is down.
+	ErrChan     chan error
+	metricsPort int
+}
+
+func NewMetricsExporter(component string) *MetricsExporter {
+	return &MetricsExporter{
+		component: component,
+		ErrChan: make(chan error),
+		metricsPort: METRICSPORT,
+	}
+}
+
+func (me *MetricsExporter) SetPort(port int) *MetricsExporter {
+	me.metricsPort = port
+	return me
+}
+
+func (me *MetricsExporter) Start() {
+	go me.serveMetrics()
+}
+
+func (me *MetricsExporter) AddRequestCounter(kind string, namespace string, action string) {
+	labels := prometheus.Labels{COMPONENT: me.component, KIND: kind, NAMESPACE: namespace, ACTION: action}
+	requestCounter.With(labels).Inc()
+}
+
+func (me *MetricsExporter) AddRequestFailureCounter(kind string, namespace string, action string) {
+	labels := prometheus.Labels{COMPONENT: me.component, KIND: kind, NAMESPACE: namespace, ACTION: action}
+	requestFailureCounter.With(labels).Inc()
+}
+
+func (me *MetricsExporter) AddRequestGauge(kind string, namespace string, action string, val float64) {
+	labels := prometheus.Labels{COMPONENT: me.component, KIND: kind, NAMESPACE: namespace, ACTION: action}
+	requestGauge.With(labels).Add(val)
+}
+
+func (me *MetricsExporter) ObserveRequestExponentialLatency(kind string, namespace string, action string, val float64) {
+	labels := prometheus.Labels{COMPONENT: me.component, KIND: kind, NAMESPACE: namespace, ACTION: action}
+	requestExponentialLatency.With(labels).Observe(val)
+}
+
+func (me *MetricsExporter) serveMetrics() {
```

Comment:
`serveMetrics()` should follow the same pattern as `server.Serve()` to directly return an error and block the calling thread if no error occurs.

In general, calling a func to do a task in a single thread or multi threads is the caller's decision. The function like `serveMetrics` or `server.Serve()` should not decide it for the caller even if it is blocking. Because in some case, the caller simply wants to be blocked on this call.

For callers which don't, it is straightforward to add an error channel and start a Goroutine. So the `Start()` and `ErrChan` add little value but bloat the API.